File: src/main/java/com/example/ticketRush/UserModule/Service/UserService.java
package com.example.ticketRush.UserModule.Service;

import com.example.ticketRush.UserModule.Entity.User;
import com.example.ticketRush.UserModule.Enum.Role;

import java.util.Optional;

public interface UserService {
    void findOrCreateUserFromKeycloak(String keycloakSubjectId, String username, String email, String fullName, Role role);

    /**
     * Tìm user bằng email hoặc username (case-insensitive).
     */
    Optional<User> findByIdentifier(String identifier);

    /**
     * Lấy danh sách users từ Keycloak và ghi ra file JSON
     */
    void exportUsersToJsonFile(String filePath);

    /**
     * Lấy danh sách users trực tiếp từ Keycloak (không qua file trung gian)
     */
    java.util.List<java.util.Map<String, Object>> getAllKeycloakUsers();

    /**
     * Tạo người dùng mới trong Keycloak
     */
    void createKeycloakUser(String username, String email, String password, String firstName, String lastName, String phoneNumber, String role, String gender, String dob);
    /**
     * Xóa người dùng trong Keycloak theo ID
     */
    void deleteKeycloakUser(String userId);

    /**
     * Cập nhật thông tin người dùng trong Keycloak
     */
    void updateKeycloakUser(String userId, String firstName, String lastName, String email, String phoneNumber, String role, String gender, String dateOfBirth);
}

File: FE/src/app/services/user-service.ts
import apiClient from "../api-client";
import { User } from "../types";

/**
 * Service xử lý các nghiệp vụ liên quan đến người dùng
 */
export const userService = {
  /**
   * Lấy danh sách tất cả người dùng từ Keycloak (qua Backend API)
   * @returns Promise<User[]>
   */
  async getUsers(): Promise<User[]> {
    try {
      const response = await apiClient.get<User[]>("/api/admin/users");
      return response.data;
    } catch (error) {
      console.error("Lỗi khi lấy danh sách user:", error);
      throw error;
    }
  },

  async createUser(data: {
    username: string;
    email: string;
    password: string;
    firstName: string;
    lastName: string;
    phoneNumber: string;
    role: string;
    gender: string;
    dateOfBirth: string;
  }): Promise<void> {
    await apiClient.post("/api/admin/users", data);
  },

  async updateUser(userId: string, data: {
    email: string;
    firstName: string;
    lastName: string;
    phoneNumber: string;
    role: string;
    gender: string;
    dateOfBirth: string;
  }): Promise<void> {
    await apiClient.put(`/api/admin/users/${userId}`, data);
  },

  async deleteUser(userId: string): Promise<void> {
    await apiClient.delete(`/api/admin/users/${userId}`);
  },
};
